### ISSUE TITLE
WebviewPanel.reveal won't reveals to `SIDE_GROUP` if the specified group does not exist

### ISSUE THREAD

**Issue by Vigilans** [microsoft/vscode#71608]
* Existing issues searched: true
* Tested under insider edition: true
* Does this issue occur when all extensions are disabled?: true
- VS Code version: Code 1.32.3 (a3db5be9b5c6ba46bb7555ec5d60178ecc2eaae4, 2019-03-14T23:43:35.476Z)
- OS version: Windows_NT x64 10.0.18362

### Steps to reproduce

1. Create a webview panel with `viewColumn` set to One
2. Call `reveal` on the panel with `viewColumn` set to Two (not exists yet)
3. The panel only reveals to Column One.

### Workaround to make it right
1. Create a webview panel with `viewColumn` set to One
2. **Execute `workbench.action.focusSecondEditorGroup` command to ensure Group Two exists**
3. Call `reveal` on the panel with `viewColumn` set to Two (exists now)
4. The panel reveals to Column Two.

### showTextDocument
Designed behavior
* When clicking `Code Now`, a doc will be called by `showTextDocument` in Column 3.

 ![demo2](https://user-images.githubusercontent.com/20227484/55531360-8d01a580-56dc-11e9-8f16-85206227a6d1.gif)
So:
  1. If I have 3 Columns, then doc will be open in Column 3.
  2. If I have 2 Columns, but Column 2 is empty, the doc will be open in Column 2.
  3. If I have 1 non-empty Column, then doc will be open in Column 2.

  This behavior is as by design.

### reveal
Designed behavior
* When **creating the webview**, the panel will be called by `reveal` to **Column 3**.
* When **clicking `Code Now`**, the panel will be called by `reveal`  to **Column 2**.

 ![demo1](https://user-images.githubusercontent.com/20227484/55424235-5def1500-55b2-11e9-84d3-4bf55879a8a4.gif)
↑ Here, if all 3 columns exists, the panel will be created to Column 3 and moved to Column 2, as by design.  

 ![demo2](https://user-images.githubusercontent.com/20227484/55424239-5e87ab80-55b2-11e9-9005-dca3286d7492.gif)
↑ Here, what happened?:
  1. If I have 2 **non-empty** Columns, and creating a new webview, the panel get revealed to **Column 2**, which **should have been in Column 3**.
  2. If I have 2 **non-empty** Columns, and clicking `Code Now`, the panel get revealed to **Column 2**, which is as by design.
  3. If I have 1 **non-empty** Columns, and creating a new webview, the panel get revealed to **Column 1**, which **should have been in Column 2**.
  4. If I have 1 **non-empty** Columns, and clicking `Code Now`, the panel get revealed to **Column 1**, which **should have been in Column 2**.

> we only allow revealing text documents in the group one greater than the current number of groups.

So, for `WebviewPanel.reveal`, it's not `one greater than`, it's `equal to`,  which is not by design.

**Comment by mjbvz** [comment]
Does this reproduce in the [latest VS Code insiders build](https://code.visualstudio.com/insiders/)?

**Comment by Vigilans** [comment]
Yes, reproduced in this version:
* VS Code version: Code - Insiders 1.33.0-insider (0ac111466ec925fd5ca180ab9893f51531e419b2, 2019-04-01T15:51:40.913Z)
* OS version: Windows_NT x64 10.0.17763

**Comment by Vigilans** [comment]
@mjbvz I'm sorry, after more tests, I think I misinterpret the problem cause.
The reason of misbehavior is that, when revealing a webview panel to column 2, and if there is no such column, then the panel won't be moved to column 2, instead it got revealed in column 1.

Here is the demo:
When all three columns exists, everything works fine.
![demo1](https://user-images.githubusercontent.com/20227484/55424235-5def1500-55b2-11e9-84d3-4bf55879a8a4.gif)
When the column is closed one by one, the stated behavior emerges.
![demo2](https://user-images.githubusercontent.com/20227484/55424239-5e87ab80-55b2-11e9-9005-dca3286d7492.gif)

The code used:
```ts
public async show(node: IProblem): Promise<void> {
    if (!this.panel) {
        this.panel = window.createWebviewPanel(
            "leetcode.preview", 
            "Preview Problem", 
            ViewColumn.Active, {
                enableScripts: true,
                enableCommandUris: true,
                enableFindWidget: true,
                retainContextWhenHidden: true,
                localResourceRoots: markdownEngine.localResourceRoots,
            });

        this.panel.webview.onDidReceiveMessage(async message => {
            switch (message.command) {
                case "ShowProblem": {
                    await commands.executeCommand("leetcode.showProblem");
                    this.panel!.reveal(ViewColumn.Two, true);
                    break;
                }
            }
        }, this, this.context.subscriptions);

        this.panel.onDidDispose(() => {
            this.panel = undefined;
        }, null, this.context.subscriptions);
    }

    const description = this.parseDescription(node);
    this.panel.webview.html = this.getWebViewContent(description);
    this.panel.title = `${node.name}: Preview`;
    this.panel.reveal(ViewColumn.Three);
}
```

**Comment by Vigilans** [comment]
@mjbvz I've built vscode locally, debugged with the first code demo provided, and finally found the cause of such behavior:

1. In the `reveal`'s RPC [entry function](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/electron-browser/mainThreadWebview.ts#L205):
 ```ts
public $reveal(handle: WebviewPanelHandle, showOptions: WebviewPanelShowOptions): void {
    const webview = this.getWebview(handle);
    if (webview.isDisposed()) {
        return;
    }

    // ↓ This line
    const targetGroup = this._editorGroupService.getGroup(
        viewColumnToEditorGroup(this._editorGroupService, showOptions.viewColumn)
    ) || this._editorGroupService.getGroup(webview.group || 0);
    
    if (targetGroup) {
        this._webviewService.revealWebview(webview, targetGroup, !!showOptions.preserveFocus);
    }
}
```
The `targetGroup` will be set to 0, the same as `webview.group`. However, the desired target group is 1.

2. In [`viewColumnToEditorGroup`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/common/shared/editor.ts#L12):
```ts
export function viewColumnToEditorGroup(editorGroupService: IEditorGroupsService, position?: EditorViewColumn): GroupIdentifier {
    if (typeof position !== 'number' || position === ACTIVE_GROUP) {
        return ACTIVE_GROUP; // prefer active group when position is undefined or passed in as such
    }

    const groups = editorGroupService.getGroups(GroupsOrder.GRID_APPEARANCE);

    let candidate = groups[position];
    if (candidate) {
        return candidate.id; // found direct match
    }

    let firstGroup = groups[0];
    if (groups.length === 1 && firstGroup.count === 0) {
        return firstGroup.id; // first editor should always open in first group independent from position provided
    }

    return SIDE_GROUP; // open to the side if group not found or we are instructed to
}
```
It will eventually return `SIDE_GROUP`.

3. In [`this._editorGroupService.getGroup`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/editorPart.ts#L260):
```ts
getGroup(identifier: GroupIdentifier): IEditorGroupView | undefined {
    return this.groupViews.get(identifier);
}
```
The `this.groupViews` is just a Map containing 1 element, whose key is 0, whereas `identifier` equals to `SIDE_GROUP`, whose value is -2. 

It will return undefined without any saying, which just means `SIDE_GROUP` does not come into any effect :(

4. So,  the `targetGroup` will just fall back to the webview panel's groups, or just the first group, further proving that `SIDE_GROUP` didn't do anything here.
```ts
const targetGroup = this._editorGroupService.getGroup(
    viewColumnToEditorGroup(this._editorGroupService, showOptions.viewColumn)
) || this._editorGroupService.getGroup(webview.group || 0);
```

5. Therefore, `this._webviewService.revealWebview` gets called with two same group, which in essence means `reveal` won't move to the group which does not exist yet.

Is this the desired behavior?

**Comment by mjbvz** [comment]
Thanks for investigating.

This behavior matches how `showTextDocument` works: we only allow revealing text documents in the group one greater than the current number of groups. So if you have no groups open, `showTextDocument` with `vscode.ViewColumn.Eight` always opens the document in group 1. If you have two groups, then `showTextDocument` with `vscode.ViewColumn.Eight` would open in group 3.

Closing this issue as by design. Let me know if I have misunderstood the problem

**Comment by Vigilans** [comment]
@mjbvz Sorry, but your statement makes sure that it is **NOT** by design.

Here is a contrast example:

### showTextDocument
Designed behavior
* When clicking `Code Now`, a doc will be called by `showTextDocument` in Column 3.

 ![demo2](https://user-images.githubusercontent.com/20227484/55531360-8d01a580-56dc-11e9-8f16-85206227a6d1.gif)
So:
  1. If I have 3 Columns, then doc will be open in Column 3.
  2. If I have 2 Columns, but Column 2 is empty, the doc will be open in Column 2.
  3. If I have 1 non-empty Column, then doc will be open in Column 2.

  This behavior is as by design.

### reveal
Designed behavior
* When **creating the webview**, the panel will be called by `reveal` to **Column 3**.
* When **clicking `Code Now`**, the panel will be called by `reveal`  to **Column 2**.

 ![demo1](https://user-images.githubusercontent.com/20227484/55424235-5def1500-55b2-11e9-84d3-4bf55879a8a4.gif)
↑ Here, if all 3 columns exists, the panel will be revealed to Column 3, as by design.  

 ![demo2](https://user-images.githubusercontent.com/20227484/55424239-5e87ab80-55b2-11e9-9005-dca3286d7492.gif)
↑ Here, what happened?:
  1. If I have 2 **non-empty** Columns, and creating a new webview, the panel get revealed to **Column 2**, which **should have been in Column 3**.
  2. If I have 2 **non-empty** Columns, and clicking `Code Now`, the panel get revealed to **Column 2**, which is as by design.
  3. If I have 1 **non-empty** Columns, and creating a new webview, the panel get revealed to **Column 1**, which **should have been in Column 2**.
  4. If I have 1 **non-empty** Columns, and clicking `Code Now`, the panel get revealed to **Column 1**, which **should have been in Column 2**.

> we only allow revealing text documents in the group one greater than the current number of groups.

So, it's not `one greater than`, it's `equal to`, just as my code analysis above suggests, which is **NOT** by design.

**Comment by Vigilans** [comment]
Is there any progress in investigating this issue?

**Comment by mjbvz** [comment]
Prs welcome

**Comment by mjbvz** [comment]
The original repo should work now. Not sure if it was fixed sometime previously or as part of dcb4759870d0ef1fc5799d0b67aa18a0da525323

The one caveat I noticed is that you must have at least one other editor open in the first group for the move to happen (or you can set `"workbench.editor.closeEmptyGroups": false`). We don't allow empty editor groups by default, so trying to move a webview that exists in its own group will simply close the existing group and then create a new one with that webview in it

Let me know if this still isn't workin in the latest insiders build

**Comment by rchiodo** [comment]
/verified

Changed the webview sample to always reveal in the second column.
